http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/monitor/GceRuntimeMonitor.scala

```diff
@@ -55,52 +63,90 @@ class GceRuntimeMonitor[F[_]: Parallel](
       startMonitoring <- nowInstant
       monitorContext = MonitorContext(startMonitoring, runtimeAndRuntimeConfig.runtime.id, traceId, action)
       _ <- Timer[F].sleep(config.initialDelay)
-      poll = googleComputeService
-        .pollOperation(googleProject, operation, config.pollingInterval, config.pollCheckMaxAttempts)
-        .compile
-        .lastOrError
-      op <- poll
+      haltWhenTrue = (Stream
+        .eval(clusterQuery.getClusterStatus(runtimeAndRuntimeConfig.runtime.id).transaction.map { newStatus =>
+          if (action == RuntimeStatus.Deleting) // Deleting is not interruptible
+            false
+          else {
+            newStatus != Some(action) && newStatus == Some(
+              RuntimeStatus.Deleting
+            ) // Interrupt Stopping if new Deleting request comes in
+          }
+        }) ++ Stream.sleep_(5 seconds)).repeat
+      _ <- if (pollCheckSupportedStatuses.contains(action))
+        F.unit
+      else F.raiseError(new Exception(s"Monitoring ${action} with pollOperation is not supported"))
+      _ <- computePollOperation
+        .pollOperation(googleProject,
+                       operation,
+                       config.pollingInterval,
+                       config.pollCheckMaxAttempts,
+                       Some(haltWhenTrue))(
+          handlePollCheckCompletion(monitorContext, runtimeAndRuntimeConfig),
+          handlePollCheckTimeout(monitorContext, runtimeAndRuntimeConfig),
+          handlePollCheckWhenInterrupted(monitorContext, runtimeAndRuntimeConfig)
```

Comment:
hmm... this monitoring functionality is getting quite complex... we're basically implementing our own event system. There does seem to be any good alternatives though